### DISCUSSION TITLE
subscribeOne vs subscribe vs unsubscribe

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#954] [Q&A] [answered]
In 0.8.0-rc2, it looks like `unsubscribe()` will unsubscribe from everything. Is that right?

Example:

```ts
client.collection('foo').subscribe(42, e=>console.log(e) )
client.collection('foo').subscribe( e=>console.log(e) )
client.collection('foo').unsubscribe() // This unsubscribes the 42 watcher as well?
```

To me it makes more sense for `subscribe()` to return an `()=>void` unsubscriber that undoes only and exactly what the corresponding `subscribe()` did. Maybe having an `unsubscribeAll()` for situations where the dev truly does want to kill all listeners.

Presently, I already have a wrapper to manage independent duplicate subscriptions to the same ID, but I'm not sure how I'd modify it to manage full table subscriptions since there doesn't seem to be any way to unsubscribe to a full table subscription without killing all listeners.

Do I understand that right?

**Comment by ganigeorgiev** [comment, marked as answer]
Calling just `client.collection('foo').unsubscribe()` will unsubscribe from all "foo" subscriptions.

Optionally you can pass only the record ids that you want to unsubscribe:
```js
client.collection('foo').unsubscribe(id1, id2, id3, ...);
```

Here is the full definition:
```js
async unsubscribe(...recordIds: Array<string>): Promise<void>
```

Additionally, I'm not sure if it is a typo but the single subscription is `subscribeOne`:
```js
client.collection('foo').subscribeOne(id, (e) => console.log(e) )
```

---

> To me it makes more sense for subscribe() to return an ()=>void...

This cannot be done because `subscribe`, `subscribeOne` and `unsubscribe` makes a server request (`POST /api/realtime`) and must return a Promise.

**Reply by ganigeorgiev** [reply]
> but I'm not sure how I'd modify it to manage full table subscriptions since there doesn't seem to be any way to unsubscribe to a full table subscription without killing all listeners

This cannot be done with the above api. As a workaround you could call the `client.realtime.unsubscribe("foo")`.

Internally all `collection().subscribe`, `collection().subscribeOne` and `collection().unsubscribe` calls `client.realtime` but map the event to a record specific type. You can check their implementation in https://github.com/pocketbase/js-sdk/blob/rc/src/services/RecordService.ts#L67-L101

**Reply by ganigeorgiev** [reply]
We could add `unsubscribeAll()` and change the behavior of `client.collection().unsubscribe()` to unsubscribe only from the collection subscription and to leave the individual ones unless id(s) are specified, but I'm not sure if it is clear enough. I'll think more on it tomorrow. Suggestions for better api are welcomed.

**Reply by ganigeorgiev** [reply]
Alternatively, we can minimize the collection realtime handlers to only 2 forms:
```js
pb.collection("foo").subscribe(topic, fn)
pb.collection("foo").unsubscribe(topic?)
```

Where topic is either a wildcard (eg. "*") or a record id. For example:

```js
// subscribe to everything within foo
pb.collection("foo").subscribe("*", fn)

// subscribe to only "foo/id"
pb.collection("foo").subscribe(id, fn)

// unsubscribe from only "*" topic
pb.collection("foo").unsubscribe("*")

// unsubscribe from only "foo/id" topic
pb.collection("foo").unsubscribe(id)

// unsubscribe from all foo subscribed topics
pb.collection("foo").unsubscribe()
```

**Reply by ganigeorgiev** [reply]
**First, what are the existing shortcomings of the current realtime handlers:**

1. We can't subscribe to the same topic multiple times, eg. the below will register only the second listener:
  
      ```js
      pb.collection("foo").subscribeOne(id, fn1);
      pb.collection("foo").subscribeOne(id, fn2); -> only fn2 will be registered
      ```

2. We can't unsubscribe only from the collection subscription while keeping the individual record subscriptions active.
      
      ```js
      pb.collection("foo").subscribe(fn);
      pb.collection("foo").subscribeOne(id, fn);
      pb.collection("foo").unsubscribe(); -> will unsubscribe from everything related to foo
      ```

**To fix 1)** we'll add support for storing multiple listeners to the same topic (this will require change in `RealtimeService.ts` but it is safe since the prop that is going to change is `private`).

**To fix 2)** without breaking already migrated v0.8 realtime code, we will:

- Use the suggested api from the previous comment (aka. we will have only `subscribe` and `unsubscribe`)

- Create a function overload for `pb.collection()subscribe(fn)`, mark it as `@deprecated` and internally call `pb.collection(foo).subscribe("*", fn)`.

- Mark `pb.collection(foo).subscribeOne(id, fn)` as `@deprecated` and internally call `pb.collection(foo).subscribe(id, fn)`.

Or at the end we will have something like this:

```js
// ---------------------------------------------------------------------------
// Soft-deprecate:
// ---------------------------------------------------------------------------

// will have a warning that it is deprecated with a suggestion to use subscribe("*", fn)
pb.collection("foo").subscribe(fn)

// will have a warning that it is deprecated with a suggestion to use subscribe(id, fn)
pb.collection("foo").subscribeOne("id", fn)

// ---------------------------------------------------------------------------
// Subscribe:
// ---------------------------------------------------------------------------

// subscribe to only "*" topic (aka. everything within foo)
 pb.collection("foo").subscribe("*", fn)

// subscribe to the same "foo/id" topic multiple times
pb.collection("foo").subscribe("id", fn1)
pb.collection("foo").subscribe("id", fn2)
const fn3Removal = await pb.collection("foo").subscribe("id", fn3)

// ---------------------------------------------------------------------------
// Unsubscribe:
// ---------------------------------------------------------------------------

// unsubscribe from all "*" topic listeners (we have only one - fn)
pb.collection("foo").unsubscribe("*")

// remove only the "foo/id" fn3 listener
fn3Removal()

// unsubscribe from all remaining "foo/id" topic listeners (fn1 and fn2)
pb.collection("foo").unsubscribe("id")

// unsubscribe from all foo listeners ("*" and any individual record ids)
pb.collection("foo").unsubscribe()
```

I think with the above we'll handle all possible cases while at the same time simplifying the api and bringing it to practically only 2 public functions:
```js
pb.collection("foo").subscribe(topic, fn): Promise<SingleUnsubscribeFunc>
pb.collection("foo").unsubscribe(topic?): Promise<void>
```

@benallfree and anyone else who stumbled on this, please let me know if you think that the above proposal is missing something. I plan to implement it before the final v0.8.0 release.

**Reply by benallfree** [reply]
I like the topic idea a lot. 

I still think the JS PocketBase client could keep a count of listeners and forward the subscribe/unsubscribe underlying command only when the first listener subscribes and the last listener unsubscribes from a topic. 

There are lots of use cases where different sections of code may want to listen to the same topic over different intervals of time. The api can still be async, but it just returns an  async unsub func that either unsubs just that one listener, or, if it is the only listener remaining, unsubs that entire topic.

**Reply by denisgurev** [reply]
If I'm understanding the example correctly you'll be able to subscribe several times to the same thing, right?
If that is implemented that would be great because currently I have a single global function that register my all other subscriptions. It workds but it's a little messy and the code would be a lot clear if it was possible to call subscribe to the same thing from different places.

**Reply by ganigeorgiev** [reply]
@benallfree, @denisgurev That's exactly what the proposal suggest. Here is the part of the above example that shows this:

```js
// subscribe to the same "foo/id" topic multiple times
pb.collection("foo").subscribe("id", fn1)
pb.collection("foo").subscribe("id", fn2)
const fn3Removal = await pb.collection("foo").subscribe("id", fn3) -> get the single fn3 removal to call later if needed
```
In the above all `fn1`, `fn2` and `fn3` are attached to the same topic.

The only thing I'm not sure yet is whether in the above the `subscribe` should return a `Promise<SingleUnsubscribeFunc>` or to change the `unsubscribe` similar to how the native JS [`removeEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener) works, aka. if you want to unsubsribe only `fn3`, you'll call instead:

```js
// remove only the fn3 listener from the "foo/id" topic
pb.collection("foo").unsubscribe("id", fn3)

// remove all listeners from the "foo/id" topic
pb.collection("foo").unsubscribe("id")

// remove all listeners from all topics in the foo collection
pb.collection("foo").unsubscribe()
```

Sometime later this evening I'll start working on the implementation and will see how it'll be best to proceed.

**Reply by benallfree** [reply]
@ganigeorgiev In my opinion, reactive programming has popularized and proven the benefits of returning an unsub func instead of requiring the dev to keep references to the original topic and callback. It also makes it easier to use a lambda callback and is more composable.

**Reply by benallfree** [reply]
I propose that the new public API can be reduced to:

```ts
async subscribe(topic:string, callback: ( e: SubscribeEvent)=>void ): Promise<UnsubscribeFunc>
```

where `topic` is a record ID or `*`. No need for a public `unsubscribe` at all. 

If you really want to offer an `unsubscribe` function, let it accept a callback reference and make sure that:
* It does the exact same thing as the unsub func
* Calling it repeatedly has no effect
* Calling it after using the unsub func has no effect
* Calling the unsub func after using `unsubscribe` has no effect
* Calling it with the wrong topic has no effect

One note/thought: internally, consider using a debounce technique so a barrage of `subscribe()` calls to different topics will be merged before the request goes over the wire.

**Reply by benallfree** [reply]
Oh please also make sure that the `subscribe` callback only fires  until the unsub func has been called, not completed. In other words, the returned unsub func immediately (synchronously) removes the callback from dispatch regardless of whether it has to await a REST request to perform an underlying unsub.

**Reply by ganigeorgiev** [reply]
@benallfree, @denisgurev I've pushed a new rc version of the [JS SDK v0.8.0-rc2](https://github.com/pocketbase/js-sdk/releases/tag/v0.8.0-rc2). Please update your dependency and share your feedback. Tomorrow after work I'll apply the same type of changes to the Dart SDK and will update PocketBase Admin "API preview" Realtime docs to reflect the new API.

**Reply by benallfree** [reply]
@ganigeorgiev Working perfectly for me so far, thank you!

**Reply by CodyBontecou** [reply]
Is the `subscribeOne` syntax still valid? 

VSCode is slashing it out, making me think it's deprecated.

<img width="331" alt="Screenshot 2022-12-23 at 8 04 49 AM" src="https://user-images.githubusercontent.com/20440899/209386843-207f4b34-51d2-42d9-a163-788a921e440a.png">

**Reply by ganigeorgiev** [reply]
@CodyBontecou No, it is deprecated. Use `pb.collection('example').subscribe('YOUR_RECORD_ID', callback)`.

**Reply by CodyBontecou** [reply]
Gotcha. What is the difference between `pb.collection('example').subscribe('YOUR_RECORD_ID', callback)` and `pb.realtime.subscribe("...", callback)`?

**Reply by ganigeorgiev** [reply]
`pb.collection('example').subscribe('YOUR_RECORD_ID', callback)` is a collection/record specific subscription. Internally it calls `pb.realtime.subscribe('example/YOUR_RECORD_ID', callback)` but also wraps the event data in a `RecordSubscription` interface.

`pb.realtime.subscribe` is usually intended to be used with custom user defined realtime subscriptions (eg. when extending PocketBase), similar to how you can call `pb.send()` to send a HTTP request to a custom API endpoint.

**Reply by CodyBontecou** [reply]
Hmm, okay. It's making more sense now.

One last thing, what is the `YOUR_RECORD_ID` in this case? Is it the `id` column?

I'm trying this subscription:
```ts
  await $pb?.collection("post").subscribe("5nq1i1bin75bxsl", async (e) => {
    await getPostList();
  });
```
<img width="427" alt="Screenshot 2022-12-23 at 8 22 00 AM" src="https://user-images.githubusercontent.com/20440899/209391195-c6c4ac8a-adff-4b36-b670-9b0c40236ffc.png">

But I do not see any response when this row is changed.

**Reply by ganigeorgiev** [reply]
@CodyBontecou Yes, it is the record id. When subscribing to a single record, make sure that the correct "View API rule" is set (by default only admins are allowed).

From the docs in https://pocketbase.io/docs/api-realtime/ 

> When you subscribe to a **single record**, the collection's **ViewRule** will be used to determine whether the subscriber has access to receive the event message.
>
> When you subscribe to an **entire collection**, the collection's **ListRule** will be used to determine whether the subscriber has access to receive the event message.

_(you can find also realtime docs and examples in the "API Preview" for your collection from the Admin UI)_

**Reply by ganigeorgiev** [reply]
@CodyBontecou You can also refer to the short service descriptions in https://github.com/pocketbase/js-sdk#realtime-handlers _(each method also comes with detailed jsdoc that you should be able to see when hover on the method in your editor)_.

**Reply by CodyBontecou** [reply]
Thank you, I appreciate the help. I was able to get it to work.

It turns out my collection is named `posts`, so when I subscribed to `post`, it wasn't returning anything haha. User error, my bad.

**Comment by benallfree** [comment]
@ganigeorgiev should both handlers get called now?

```ts
const unsub1 = client.collection('orders').subscribe('*', handler1)
const unsub2 = client.collection('orders').subscribe('*', handler2)
```

**Reply by benallfree** [reply]
I think there is a bug, `this.eventSource.addEventListener(topic, listener);` needs to be called for every add, right?

**Reply by ganigeorgiev** [reply]
@benallfree, Yes, both `handler1` and `handler2` should be called but I think we already do this in https://github.com/pocketbase/js-sdk/blob/rc/src/services/RealtimeService.ts#L205-L217.

Also note that I've pushed some minor changes to the rc branch but I haven't submitted a new release yet (probably will be available with the final v0.8.0)

btw. you have to `await` the subscribe() call in order to get the `UnsubscribeFunc` because the subscription may fail, aka.:
```js
const unsub1 = await client.collection('orders').subscribe('*', handler1)
const unsub2 = await client.collection('orders').subscribe('*', handler2)
```

**Reply by ganigeorgiev** [reply]
@benallfree I've just tested with JS SDK v0.8.0-rc2 and it works as expected - both handlers are fired on `orders` change. Make sure that you have "List API rule" access for your collection when you subscribe to the wildcard topic.

**Reply by benallfree** [reply]
@ganigeorgiev Can you please take another look? As far as i can see, `await this.submitSubscriptions()` is only called when the first listener is added:

https://github.com/pocketbase/js-sdk/blob/7956bfe992cea931a120f039e0d50c479e910e84/src/services/RealtimeService.ts#L45
```ts
        } else if (this.clientId && this.subscriptions[topic].length === 1) {
            // send the updated subscriptions (if it is the first for the topic)
            await this.submitSubscriptions();
        }
```

`submitSubscriptions` is the only place `addAllSubscriptionListeners` is called, which is the only place where `this.eventSource.addEventListener` is called. https://github.com/pocketbase/js-sdk/blob/7956bfe992cea931a120f039e0d50c479e910e84/src/services/RealtimeService.ts#L214

This is consistent with my experience which is that only the FIRST handler is actually called because subsequent `subscribe()` calls do not actually connect the callback with `addEventListener`.

**Reply by ganigeorgiev** [reply]
@benallfree Actually, you are right. Currently it works only if you register all subscriptions before the first PB_CONNECT. I'll update it shortly and will push a new release.

**Reply by ganigeorgiev** [reply]
@benallfree It should be fixed in the latest [JS SDK v0.8.0-rc3 release](https://github.com/pocketbase/js-sdk/releases/tag/v0.8.0-rc3).

**Reply by benallfree** [reply]
Perfect, works now!

**Reply by benallfree** [reply]
Hey @ganigeorgiev here is something to consider for unsubscribe:

In the pockethost dashboard, it subscribes to realtime updates for each of the user's instances individually. https://github.com/benallfree/pockethost/blob/master/packages/pockethost.io/src/routes/dashboard/%2Bpage.svelte#L50. When the component unmounts, it causes one unsubscribe per instance.

On https://github.com/pocketbase/js-sdk/blob/rc/src/services/RealtimeService.ts#L88, it calls `submitSubscriptions()` which causes a `client.send` unconditionally. This causes 2 issues:

1. Essentially a debounce situation where 20 network requests are made instead of what could be more efficiently done as 1.
2. Potential needless network requests because it really only needs to send a realtime update when `Object.keys(this.subscriptions)` changes. As far as I can see, it blindly sends an update even when `Object.keys(this.subscriptions)` is unchanged.

Also, and importantly :), I think there is a bug here https://github.com/pocketbase/js-sdk/blob/rc/src/services/RealtimeService.ts#L149 where a `if(this.subscriptions[topic].length===0) delete this.subscriptions[topic]` is needed. Right now, individual unsubscribes fail to delete an empty topic so the realtime update actually keeps the topic active.

For example, after the dashboard unmount, all the `instances/<id>` topics are unsubscribed, but they are still sent in the realtime update that was supposed to drop them:

```
{
	"clientId": "KSIvl7y2p2nOOPX9TJBK6F649AdF1cmiX9z4B3X9",
	"subscriptions": [
		"instances/q6gg6yqiqojpju8",
		"instances/dvb32cc5g1ip79e",
		"instances/kpa3osdil1cvcfv",
		"instances/tu2m0lqopflogw5",
		"instances/98wim7ld734yeye",
		"instances/i27d7xdfhrmfo7b",
		"instances/tmkpemjgye745dg",
		"rpc/h5kba4k71jt4bw6",
		"instances/nqgvvy20opwo9oy",
		"backups"
	]
}
```

**Reply by ganigeorgiev** [reply]
@benallfree I'll fix the second issue when the topic doesn't have any subscriptions, but I'm not sure that I understand the first one. Could you provide a simplified example when do you experience the 20 network requests (also note that even if there are 20 requests only 1 should resolve due to the autocancellation)?

**Reply by benallfree** [reply]
Sorry yes, here's a much simpler pattern.

I do have autocancellation disabled because there are many cases in the frontend where concurrent fetches (fetching different `instances` records for example) are desirable and awaiting them serially causes noticeable ux wait times. 

```ts
client.beforeSend = (url: string, reqConfig: { [_: string]: any }) => {
  delete reqConfig.signal
  return reqConfig
}

const unsubs = await Promise.all([
  client.collection('instances').subscribe(id1, myCallback),
  client.collection('instances').subscribe(id2, myCallback),
  client.collection('instances').subscribe(id3, myCallback),
  client.collection('instances').subscribe(id4, myCallback),
  client.collection('instances').subscribe(id5, myCallback),
])

unsubs.forEach(u=>u()) // This will cause 5 network requests 
```

I think a debounced mutex makes sense for subscription updates. There really should only be one `/realtime` POST happening at any time, regardless of whether autocancellation is enabled, and it should always be sending the latest `Object.keys(this.subscriptions)`.

Here is an implementation (untested) that would limit requests to one a at a time, debounce, and queue pending updates while in-flight:

```ts
const limiter = new Bottleneck({ maxConcurrent: 1 })
async function submitSubscriptions(): Promise<boolean> {
  if (!this.clientId) {
    return false
  }

  // optimistic update
  this.addAllSubscriptionListeners()

  /**
   * If there is already a queued subscription update, return the
   * existing promise. In the next event loop, it will be executed.
   * This allows multiple unsubs to be called synchronously without
   * causing multiple network requests even when autocancellation
   * is disabled.
   */
  if (this.queuedUpdatePromise) return this.queuedUpdatePromise

  /**
   * We can only get here if there is no queued update. Because
   * there still might be an update currently in-flight, we use
   * the limiter set to maxConcurrent: 1. This new promise will 
   * not execute until the current in-flight update (if any) resolves.
   */
  this.queuedUpdatePromise = limiter.schedule(
    new Promise((resolve) => {
      setTimeout(() => {
        this.queuedUpdatePromise = undefined // Allow another update to be queued if subscriptions change after this point
        this.client
          .send('/api/realtime', {
            method: 'POST',
            body: {
              clientId: this.clientId,
              subscriptions: Object.keys(this.subscriptions),
            },
            params: {
              $cancelKey: 'realtime_subscriptions_' + this.clientId,
            },
          })
          .then(() => {
            /**
             * Instead of returning `true`, now we `resolve(true)`
             */
            resolve(true) 
          })
          .catch((err) => {
            if (err?.isAbort) {
              return true // silently ignore aborted pending requests
            }
            throw err
          })
      }, 0)
    })
  )

  return this.queuedUpdatePromise
}
```

What do you think? I can't think of a use case where stale subscription updates are ever desired.

*Side note: Making the returned unsub async is also worth considering. `subscribe(topic, callback): Promise< ()=>Promise<void> >`. They are currently a bit misleading because there is an underlying async operation and no way to catch or await it.*

**Reply by ganigeorgiev** [reply]
I'll have to think a little more on this. I don't like the above because it complicates too much the RealtimeService code (which is already messy) and produces similar results to autocancellation.
I'll consider refactoring it after v0.8.0 and eventually add tests because currently it is the only service in the JS SDK that it is not unit tested and it shows.

--- 

> Side note: Making the returned unsub async is also worth considering. 

I think it is already async, the current `subscribe` return type is:

```
async subscribe(topic: string, callback: (data: any) => void): Promise<UnsubscribeFunc>

type UnsubscribeFunc = () => Promise<void>
```

**Reply by benallfree** [reply]
Ah I missed that, you're right. 

Ok so what I'm hearing is there is currently no way to prevent duplicate network requests and possible subscription race conditions with autocancellation disabled? I don't think I can fix this in my user code because unsubscribe calls `client.send` synchronously. Any thoughts on how best to manage this for now?

**Reply by ganigeorgiev** [reply]
> Ok so what I'm hearing is there is currently no way to prevent duplicate network requests and possible subscription race conditions with autocancellation disabled?

In the above example where you call `unsubs.forEach(u=>u())`, there are no race conditions. All subscribe/unsubscribe operations acts "optimistically", aka. they first do changes on the `this.subscriptions` prop and then make the submit call to the server (this is to ensure that even on EventSource reconnect the latest user subscriptions will be submitted). Or in other words, in the worst case scenario, with auto cancellation disabled, you may end up sending the same request multiple time, but they all should be with the latest subscriptions state.

The easiest workaround I can think of at the moment would be to enable autocancellation only for the realtime service by checking the `url` for `/api/realtime` path and skip the `signal` deletion.

**Reply by benallfree** [reply]
I’ll try that. If that doesn’t work, maybe I can instantiate a second instance of the client and handle all the realtime stuff in that separate instance.

About the race conditions, if the subscriber topics change while updates are in flight, then things could get out of sync depending upon how network requests resolve. Of course, this only is a problem when auto cancellation is disabled.

**Reply by ganigeorgiev** [reply]
@benallfree The bug with the unnecessary topics submission when there are no listeners should be fixed in the latest [JS SDK v0.8.0-rc4](https://github.com/pocketbase/js-sdk/releases/tag/v0.8.0-rc4).

**Reply by benallfree** [reply]
@ganigeorgiev Thank you! What do you think about `subscribe(topicOrTopics: string | string[], ...)`? That could be a useful means of (un)subscribing a lot all at once. That said, your fix plus turning auto cancellation back on seems to be working, so I'm moving on :)

**Reply by ganigeorgiev** [reply]
@benallfree I actually thought of this initially, but decided to keep it for now only accepting single topic so that the api is simpler and as close as possible to the [Dart SDK](https://github.com/pocketbase/dart-sdk) (Dart doesn't have union types or variadic functions). Depending on the users demand for this, I may reconsider it in the future.

**Comment by ninakylle02** [comment]


**Reply by ganigeorgiev** [reply]
Please don't spam the discussion.

**Comment by ninakylle02** [comment]
Ok, thanks.

**Comment by ninakylle02** [comment]
What do you mean?

**Comment by ninakylle02** [comment]
What do you mean?
Do you not trust me tho?
